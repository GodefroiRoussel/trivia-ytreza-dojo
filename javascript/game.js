exports = typeof window !== "undefined" && window !== null ? window : global;

exports.Game = function(out) {
  var players          = new Array();
  var places           = new Array(6);
  var purses           = new Array(6);
  var inPenaltyBox     = new Array(6);

  var popQuestions     = new Array();
  var scienceQuestions = new Array();
  var sportsQuestions  = new Array();
  var rockQuestions    = new Array();

  var currentPlayer    = 0;
  var isGettingOutOfPenaltyBox = false;

  var didPlayerWin = function(){
    return !(purses[currentPlayer] == 6)
  };

  var currentCategory = function(){
    if(places[currentPlayer] == 0)
      return 'Pop';
    if(places[currentPlayer] == 4)
      return 'Pop';
    if(places[currentPlayer] == 8)
      return 'Pop';
    if(places[currentPlayer] == 1)
      return 'Science';
    if(places[currentPlayer] == 5)
      return 'Science';
    if(places[currentPlayer] == 9)
      return 'Science';
    if(places[currentPlayer] == 2)
      return 'Sports';
    if(places[currentPlayer] == 6)
      return 'Sports';
    if(places[currentPlayer] == 10)
      return 'Sports';
    return 'Rock';
  };

  this.createRockQuestion = function(index){
    return "Rock Question "+index;
  };

  for(var i = 0; i < 50; i++){
    popQuestions.push("Pop Question "+i);
    scienceQuestions.push("Science Question "+i);
    sportsQuestions.push("Sports Question "+i);
    rockQuestions.push(this.createRockQuestion(i));
  };

  this.isPlayable = function(howManyPlayers){
    return howManyPlayers >= 2;
  };

  this.add = function(playerName){
    players.push(playerName);
    places[this.howManyPlayers() - 1] = 0;
    purses[this.howManyPlayers() - 1] = 0;
    inPenaltyBox[this.howManyPlayers() - 1] = false;

    out.log(playerName + " was added");
    out.log("They are player number " + players.length);

    return true;
  };

  this.howManyPlayers = function(){
    return players.length;
  };


  var askQuestion = function(){
    if(currentCategory() == 'Pop')
      out.log(popQuestions.shift());
    if(currentCategory() == 'Science')
      out.log(scienceQuestions.shift());
    if(currentCategory() == 'Sports')
      out.log(sportsQuestions.shift());
    if(currentCategory() == 'Rock')
      out.log(rockQuestions.shift());
  };

  this.roll = function(roll){
    out.log(players[currentPlayer] + " is the current player");
    out.log("They have rolled a " + roll);

    if(inPenaltyBox[currentPlayer]){
      if(roll % 2 != 0){
        isGettingOutOfPenaltyBox = true;

        out.log(players[currentPlayer] + " is getting out of the penalty box");
        places[currentPlayer] = places[currentPlayer] + roll;
        if(places[currentPlayer] > 11){
          places[currentPlayer] = places[currentPlayer] - 12;
        }

        out.log(players[currentPlayer] + "'s new location is " + places[currentPlayer]);
        out.log("The category is " + currentCategory());
        askQuestion();
      }else{
        out.log(players[currentPlayer] + " is not getting out of the penalty box");
        isGettingOutOfPenaltyBox = false;
      }
    }else{

      places[currentPlayer] = places[currentPlayer] + roll;
      if(places[currentPlayer] > 11){
        places[currentPlayer] = places[currentPlayer] - 12;
      }

      out.log(players[currentPlayer] + "'s new location is " + places[currentPlayer]);
      out.log("The category is " + currentCategory());
      askQuestion();
    }
  };

  this.wasCorrectlyAnswered = function(){
    if(inPenaltyBox[currentPlayer]){
      if(isGettingOutOfPenaltyBox){
        out.log('Answer was correct!!!!');
        purses[currentPlayer] += 1;
        out.log(players[currentPlayer] + " now has " +
                    purses[currentPlayer]  + " Gold Coins.");

        var winner = didPlayerWin();
        currentPlayer += 1;
        if(currentPlayer == players.length)
          currentPlayer = 0;

        return winner;
      }else{
        currentPlayer += 1;
        if(currentPlayer == players.length)
          currentPlayer = 0;
        return true;
      }



    }else{

      out.log("Answer was correct!!!!");

      purses[currentPlayer] += 1;
      out.log(players[currentPlayer] + " now has " +
                  purses[currentPlayer]  + " Gold Coins.");

      var winner = didPlayerWin();

      currentPlayer += 1;
      if(currentPlayer == players.length)
        currentPlayer = 0;

      return winner;
    }
  };

  this.wrongAnswer = function(){
		out.log('Question was incorrectly answered');
		out.log(players[currentPlayer] + " was sent to the penalty box");
		inPenaltyBox[currentPlayer] = true;

    currentPlayer += 1;
    if(currentPlayer == players.length)
      currentPlayer = 0;
		return true;
  };
};

exports.start = function(out, numberGenerator) {
  var notAWinner = false;

  var game = new Game(out);
  game.add('Chet');
  game.add('Pat');
  game.add('Sue');
  do {
  
    game.roll(Math.floor(Math.random() * 6) + 1);
  
    if (Math.floor(Math.random() * 10) == 7) {
      notAWinner = game.wrongAnswer();
    } else {
      notAWinner = game.wasCorrectlyAnswered();
    }
  
  } while (notAWinner);
}
