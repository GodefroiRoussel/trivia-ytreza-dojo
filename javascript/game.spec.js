require('./game.js');

describe("The test environment", function() {
  it("should pass", function() {
    expect(true).toBe(true);
  });

  it("should access game", function() {
    expect(Game).toBeDefined();
    expect(Game).toMatchSnapshot()
    let textGlobal
    const out = {
      log: (text) => {
        textGlobal += text
      }
    }
    start(out)
    expect(textGlobal).toMatchSnapshot()
  });
});

describe("Your specs...", function() {
  // it ...
});
