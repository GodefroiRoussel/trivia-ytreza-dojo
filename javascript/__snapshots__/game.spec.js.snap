// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The test environment should access game 1`] = `[Function]`;

exports[`The test environment should access game 2`] = `"undefinedChet was addedThey are player number 1Pat was addedThey are player number 2Sue was addedThey are player number 3Chet is the current playerThey have rolled a 1Chet's new location is 1The category is ScienceScience Question 0Question was incorrectly answeredChet was sent to the penalty boxPat is the current playerThey have rolled a 4Pat's new location is 4The category is PopPop Question 0Answer was correct!!!!Pat now has 1 Gold Coins.Sue is the current playerThey have rolled a 1Sue's new location is 1The category is ScienceScience Question 1Answer was correct!!!!Sue now has 1 Gold Coins.Chet is the current playerThey have rolled a 3Chet is getting out of the penalty boxChet's new location is 4The category is PopPop Question 1Answer was correct!!!!Chet now has 1 Gold Coins.Pat is the current playerThey have rolled a 2Pat's new location is 6The category is SportsSports Question 0Answer was correct!!!!Pat now has 2 Gold Coins.Sue is the current playerThey have rolled a 1Sue's new location is 2The category is SportsSports Question 1Answer was correct!!!!Sue now has 2 Gold Coins.Chet is the current playerThey have rolled a 6Chet is not getting out of the penalty boxPat is the current playerThey have rolled a 6Pat's new location is 0The category is PopPop Question 2Answer was correct!!!!Pat now has 3 Gold Coins.Sue is the current playerThey have rolled a 1Sue's new location is 3The category is RockRock Question 0Answer was correct!!!!Sue now has 3 Gold Coins.Chet is the current playerThey have rolled a 6Chet is not getting out of the penalty boxPat is the current playerThey have rolled a 5Pat's new location is 5The category is ScienceScience Question 2Answer was correct!!!!Pat now has 4 Gold Coins.Sue is the current playerThey have rolled a 2Sue's new location is 5The category is ScienceScience Question 3Answer was correct!!!!Sue now has 4 Gold Coins.Chet is the current playerThey have rolled a 1Chet is getting out of the penalty boxChet's new location is 5The category is ScienceScience Question 4Answer was correct!!!!Chet now has 2 Gold Coins.Pat is the current playerThey have rolled a 1Pat's new location is 6The category is SportsSports Question 2Answer was correct!!!!Pat now has 5 Gold Coins.Sue is the current playerThey have rolled a 3Sue's new location is 8The category is PopPop Question 3Answer was correct!!!!Sue now has 5 Gold Coins.Chet is the current playerThey have rolled a 5Chet is getting out of the penalty boxChet's new location is 10The category is SportsSports Question 3Answer was correct!!!!Chet now has 3 Gold Coins.Pat is the current playerThey have rolled a 3Pat's new location is 9The category is ScienceScience Question 5Question was incorrectly answeredPat was sent to the penalty boxSue is the current playerThey have rolled a 1Sue's new location is 9The category is ScienceScience Question 6Answer was correct!!!!Sue now has 6 Gold Coins."`;
